examples/nlp/intent_detection_slot_tagging/joint_intent_slot_infer.py

```diff
@@ -94,38 +108,72 @@
     checkpoint_dir=args.checkpoint_dir,
 )
 
-
-def concatenate(lists):
-    return np.concatenate([t.cpu() for t in lists])
-
-
-def get_preds(logits):
-    return np.argmax(logits, 1)
-
-
-intent_logits, slot_logits, loss_mask, subtokens_mask, intents, slot_labels = [
+# --- analyse of the results ---
+intent_logits, slot_logits, loss_mask, subtokens_mask, intent_labels, slot_labels_unmasked = [
     concatenate(tensors) for tensors in evaluated_tensors
 ]
 
-pred_intents = np.argmax(intent_logits, 1)
-logging.info('Intent prediction results')
-
-intents = np.asarray(intents)
-pred_intents = np.asarray(pred_intents)
-intent_accuracy = sum(intents == pred_intents) / len(pred_intents)
-logging.info(f'Intent accuracy: {intent_accuracy}')
-logging.info(classification_report(intents, pred_intents))
-
-slot_preds = np.argmax(slot_logits, axis=2)
-slot_preds_list, slot_labels_list = [], []
+# slot accuracies
+logging.info('Slot Prediction Results:')
+slot_preds_unmasked = np.argmax(slot_logits, axis=2)
 subtokens_mask = subtokens_mask > 0.5
-for i, sp in enumerate(slot_preds):
-    slot_preds_list.extend(list(slot_preds[i][subtokens_mask[i]]))
-    slot_labels_list.extend(list(slot_labels[i][subtokens_mask[i]]))
-
-logging.info('Slot prediction results')
-slot_labels_list = np.asarray(slot_labels_list)
-slot_preds_list = np.asarray(slot_preds_list)
-slot_accuracy = sum(slot_labels_list == slot_preds_list) / len(slot_labels_list)
-logging.info(f'Slot accuracy: {slot_accuracy}')
-logging.info(classification_report(slot_labels_list, slot_preds_list))
+slot_labels = slot_labels_unmasked[subtokens_mask]
+slot_preds = slot_preds_unmasked[subtokens_mask]
+slot_accuracy = np.mean(slot_labels == slot_preds)
+logging.info(f'Slot Accuracy: {slot_accuracy}')
+f1_scores = get_f1_scores(slot_labels, slot_preds, average_modes=['weighted', 'macro', 'micro'])
+for k, v in f1_scores.items():
+    logging.info(f'{k}: {v}')
+
+logging.info(f'\n {get_classification_report(slot_labels, slot_preds, label_ids=data_desc.slots_label_ids)}')
+
+# intent accuracies
+logging.info('Intent Prediction Results:')
+intent_preds = np.asarray(np.argmax(intent_logits, 1))
+intent_labels = np.asarray(intent_labels)
+intent_accuracy = np.mean(intent_labels == intent_preds)
+logging.info(f'Intent Accuracy: {intent_accuracy}')
+f1_scores = get_f1_scores(intent_labels, intent_preds, average_modes=['weighted', 'macro', 'micro'])
+for k, v in f1_scores.items():
+    logging.info(f'{k}: {v}')
+
+logging.info(f'\n {get_classification_report(intent_labels, intent_preds, label_ids=data_desc.intents_label_ids)}')
+
+# print queries with wrong intent:
+queries = open(f'{data_desc.data_dir}/{args.eval_file_prefix}.tsv', 'r').readlines()[1:]
+intent_dict = open(data_desc.intent_dict_file, 'r').read().splitlines()
```

Comment:
Can you read dictionary from the data_desc instead of reading from file?

Reply:
I tried to do it. But it is in different format from one that I am using. It is stored there as a dictionary with name as a key and index as a value. And use it just as a list where I can access it by intent/slot index.